jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt

```diff
@@ -24,60 +31,47 @@ import com.jetbrains.rd.util.put
 import com.jetbrains.rider.model.debuggerWorker.DotNetDebuggerSessionModel
 import com.jetbrains.rider.model.debuggerWorkerConnectionHelperModel
 import com.jetbrains.rider.projectView.solution
 import com.jetbrains.rider.run.IDebuggerOutputListener
 import com.jetbrains.rider.run.bindToSettings
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
 import org.jetbrains.concurrency.AsyncPromise
 import org.jetbrains.concurrency.Promise
 import software.amazon.awssdk.services.lambda.model.Runtime
-import software.aws.toolkits.core.utils.error
-import software.aws.toolkits.core.utils.getLogger
-import software.aws.toolkits.core.utils.trace
 import software.aws.toolkits.jetbrains.services.lambda.execution.local.SamDebugSupport
 import software.aws.toolkits.jetbrains.services.lambda.execution.local.SamRunningState
+import software.aws.toolkits.jetbrains.utils.ApplicationThreadPoolScope
 import software.aws.toolkits.jetbrains.utils.DotNetDebuggerUtils
+import software.aws.toolkits.jetbrains.utils.getCoroutineUiContext
+import software.aws.toolkits.jetbrains.utils.spinUntilResult
 import software.aws.toolkits.resources.message
-import java.io.File
-import java.io.OutputStream
 import java.net.InetAddress
+import java.time.Duration
 import java.util.Timer
 import kotlin.concurrent.schedule
 
 /**
  * Rider uses it's own DebuggerWorker process that run under .NET with extra parameters to configure:
- *    - path to runtime
- *    - path to DLL
  *    - debugger mode (client/server)
  *    - frontend port
  *    - backend port
  *
- * Debugger is launched via SAM CLI command: '$ sam local invoke' and pass debugger parameters --debugger-path and --debugger-args.
- * Under SAM CLI this command is translated into:
- *    - volume that is mounted in Docker to launch debugger (--debugger-path).
- *      This volume is set to path with debugger process used in Rider - DebuggerWorker - inside Rider SDK.
- *    - additional debugger arguments that are applied when launching debugger inside Docker (--debugger-args):
- *      '$ dotnet <debug_args_list> MockBootstraps.dll --debugger-spin-wait'
- *
- * We use <debugger_args_list> to inject a simple custom dotnet core application 'JetBrains.Rider.Debugger.Launcher'
- * that run Rider's debugger. This app starts Rider's debugger under Mono runtime that is bundled with Rider SDK (runtime.sh).
- * The app read arguments and start a Rider's DebuggerWorker process with ports and compiled lambda DLL - MockBootstraps.dll
- *
- * 'JetBrains.Rider.Debugger.Launcher' app is hosted on github repo - https://github.com/JetBrains/JetBrains.Rider.Debugger.Launcher.
- * The app is packed to NuGet package and is accessible in Rider SDK in '<SDKPath>/lib/ReSharperHost/' directory.
+ * Workflow:
+ * 1. Find the correct container by filtering `doker ps` for one of the published port of the debugger
+ * 2. Use `docker exec` and pgrep to locate the dotnet PID of the runtime.
```

Comment:
are we guaranteed to have pgrep? iirc it's non-standard, even if though it's frequently available

Reply:
Good question... I'll see what I can find out

Reply:
If we can't, we will go with this beauty:

```
find /proc -mindepth 2 -maxdepth 2 -name exe -exec ls -l {} \; 2>/dev/null | grep dotnet | sed -n 's/.*\/proc\/\(.*\)\/exe.*/\1/p'
```